File: OriginalCsFiles/GachaMachine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Cysharp.Threading.Tasks;
using DG.Tweening;
using UnityEngine;

namespace GachaPerformance.Scripts
{
    public class GachaMachine : MonoBehaviour
    {
        /*
         * Machine構造レイヤーメモ。上のものほど前面になる。
         * - GachaMachineFront.svg - 304 x 456 px, origin: bottom, position: 0, 0 px
         * - GachaMachineKnob.svg - 80 x 80 px, origin: center, position: 0, 132 px
         * - この階層に内部ボールを16個配置
         * - GachaMachine.svg - 304 x 456 px, origin: bottom, position: 0, 0 px
         */
        
        [SerializeField] private List<GachaInsideBall> _gachaBalls = new List<GachaInsideBall>(); // 16個。
        [SerializeField] private List<Color> _ballColors = new List<Color>();
        /*
         * Color 一覧：
         * - #E60000
         * - #F27900
         * - #F2DE00
         * - #24D900
         * - #00AAFF
         * - #C000E6
         */
        [SerializeField] private SpriteRenderer _knob; // GachaMachineKnob.svg
        private readonly Vector2 _glassOrigin = new Vector2(0.0f, 3.0f); // ガチャ内部ボールの領域の中心座標
        private const float ContainerRadius = 1.0f;

        public void Initialize()
        {
            gameObject.SetActive(true);
            transform.position = new Vector3(0.0f, 5.0f, 0.0f); // 初期位置。画面上側の見えない位置。
            transform.rotation = Quaternion.identity;
            transform.localScale = new Vector3(1.0f, 1.2f, 1.0f);
            _knob.transform.rotation = Quaternion.identity;
            
            // 内部ガチャボールの初期設定
            var ballPositions = GenerateBallPositions(_glassOrigin, ContainerRadius, 0.32f, 64);
            for (int i = 0; i < _gachaBalls.Count; i++)
            {
                var ball = _gachaBalls[i];
                ball.transform.localPosition = new Vector3(ballPositions[i].x, ballPositions[i].y, ball.transform.position.z);
                ball.SetRandomAngle();
                var randomColor = _ballColors[UnityEngine.Random.Range(0, _ballColors.Count)];
                ball.SetColor(randomColor);
            }
        }
        
        public async UniTask FallAsync(float duration, CancellationToken ct)
        {
            var moveTask = transform.DOLocalMove(new Vector3(0.0f, -2.24f, 0.0f), duration)
                .SetEase(Ease.InQuad)
                .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: ct);
            var machineScaleTask = transform.DOScale(Vector3.one, duration)
                .SetEase(Ease.InQuad)
                .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: ct);
            
            await UniTask.WhenAll(moveTask, machineScaleTask);
        }
        
        public async UniTask LandAsync(CancellationToken ct)
        {
            var shakeTask = ShakingAnimation();
            var reboundTask = ReboundAnimation();
            await UniTask.WhenAll(shakeTask, reboundTask);
            
            return;
            
            async UniTask ShakingAnimation()
            {
                await transform.DOShakePosition(0.4f, new Vector3(0.2f, 0.0f, 0.0f), 30, 90)
                    .SetEase(Ease.Linear)
                    .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: ct);
            }
            
            async UniTask ReboundAnimation()
            {
                await transform.DOScale(new Vector3(1.1f, 0.8f, 1.0f), 0.125f)
                    .SetEase(Ease.OutQuad)
                    .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: ct);
                await transform.DOScale(new Vector3(0.95f, 1.05f, 1.0f),0.125f)
                    .SetEase(Ease.OutQuad)
                    .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: ct);
                await transform.DOScale(Vector3.one, 0.125f)
                    .SetEase(Ease.OutQuad)
                    .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: ct);
            }
        }
        
        public async UniTask BeatLoopAsync(CancellationToken ct)
        {
            try
            {
                var beatScale = new Vector3(1.05f, 0.95f, 1.0f);
                while (!ct.IsCancellationRequested)
                {
                    transform.localScale = beatScale;
                    await transform.DOScale(Vector3.one, 0.5f).SetEase(Ease.OutQuad)
                        .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: ct);
                }
            }
            finally
            {
                transform.localScale = Vector3.one;
            }
        }
        
        public async UniTask RotateKnobAsync(float duration, CancellationToken ct)
        {
            await _knob.transform.DORotate(new Vector3(0, 0, -90), duration * 0.4f)
                .SetEase(Ease.OutQuad)
                .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: ct);
            
            await UniTask.Delay(TimeSpan.FromSeconds(duration * 0.2f), cancellationToken: ct);
            
            await _knob.transform.DORotate(new Vector3(0, 0, -180), duration * 0.4f)
                .SetEase(Ease.OutQuad)
                .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: ct);
            
            // ノブの回転をリセットしておく
            _knob.transform.rotation = Quaternion.identity;
        }

        public async UniTask LotteryAsync(float duration, CancellationToken ct)
        {
            // 抽選アニメーション。
            // すべてのボールをランダムな方向に弾き飛ばしまくる

            var lotteryAnimationCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            var bounceTasks = _gachaBalls.Select(ball =>
                ball.BounceLoopAsync(_glassOrigin, ContainerRadius, 12.0f, lotteryAnimationCts.Token));
            var rotateTasks = _gachaBalls.Select(ball => ball.RotateLoopAsync(360.0f, lotteryAnimationCts.Token));
            transform.rotation = Quaternion.Euler(0.0f, 0.0f, -2.0f);
            var shakeTask = transform.DORotate(new Vector3(0.0f, 0.0f, 2.0f), 0.05f)
                .SetEase(Ease.Linear)
                .SetLoops(-1, LoopType.Yoyo)
                .ToUniTask(TweenCancelBehaviour.KillAndCancelAwait, cancellationToken: lotteryAnimationCts.Token);
            // すべてのtasksが終わるか、タイムアウト
            var timeoutTask = UniTask.Delay(TimeSpan.FromSeconds(duration), ignoreTimeScale: true, cancellationToken: lotteryAnimationCts.Token);
            await UniTask.WhenAny(UniTask.WhenAll(bounceTasks), UniTask.WhenAll(rotateTasks), shakeTask, timeoutTask);
            lotteryAnimationCts.Cancel();
        }

        public void Hide()
        {
            gameObject.SetActive(false);
        }

        /// <summary>
        /// 円形領域内で、各点間がminDistance以上離れている点群を生成し、
        /// Y座標が低い順にソートしたリストを返す。
        /// </summary>
        /// <param name="center">円の中心</param>
        /// <param name="containerRadius">円の半径</param>
        /// <param name="minDistance">各点間の最低距離</param>
        /// <param name="desiredCount">生成する点数（上限）</param>
        /// <param name="k">Poisson Disk Samplingで各点周辺に試行する候補数</param>
        /// <param name="seed">乱数シード（任意）</param>
        /// <returns>充填順（下側から）の点リスト</returns>
        private List<Vector2> GenerateBallPositions(Vector2 center, float containerRadius, float minDistance, int desiredCount, int k = 30, int? seed = null) 
        {
            /*
             * 流れ：
             * - 円周上にランダムな点を生成（この座標がないと容器内に重力にしたがって張り付いた感じが得られない）
             * - 容器内のランダムな座標に点を生成
             * -- 点同士の最低距離を確保する（ガチャボール自体のサイズを表現する）
             * - Y座標が低い順にソート
             */
            
            var rand = seed.HasValue ? new System.Random(seed.Value) : new System.Random();
            var points = new List<Vector2>();
            var activeList = new List<Vector2>();

            // 円周上のランダムな点を生成
            const int targetCircumferencePoints = 32; // 円周上の点の最大数
            int validCircumferencePoints = 0;
            int attempts = 0;
            const int maxAttempts = targetCircumferencePoints * 8; // 無限ループ回避のための上限
            while (validCircumferencePoints < targetCircumferencePoints && attempts < maxAttempts) {
                attempts++;
                float angle = (float)(rand.NextDouble() * 2 * Mathf.PI);
                var candidate = center + new Vector2(containerRadius * Mathf.Cos(angle), containerRadius * Mathf.Sin(angle));
                // 既存の候補との距離がminDistance以上かチェック
                if (!points.All(p => Vector2.Distance(candidate, p) >= minDistance)) continue;
                points.Add(candidate);
                activeList.Add(candidate);
                validCircumferencePoints++;
            }
            
            // Poisson Disk Samplingで内部の点を生成
            while (activeList.Count > 0 && points.Count < desiredCount) {
                int index = rand.Next(activeList.Count);
                var point = activeList[index];
                bool candidateAccepted = false;

                // 周辺にk個の候補点を生成
                for (int i = 0; i < k; i++) {
                    float angle = (float)(rand.NextDouble() * 2 * Mathf.PI);
                    float distance = minDistance * (1 + (float)rand.NextDouble());
                    var candidate = point + new Vector2(Mathf.Cos(angle), Mathf.Sin(angle)) * distance;

                    // 候補点が円内にあるかチェック
                    if (Vector2.Distance(candidate, center) > containerRadius)
                        continue;

                    // 既存の点すべてとの距離がminDistance以上かチェック
                    bool valid = points.All(p => !(Vector2.Distance(candidate, p) < minDistance));
                    if (!valid) continue;
                    points.Add(candidate);
                    activeList.Add(candidate);
                    candidateAccepted = true;
                    break;
                }
                if (!candidateAccepted) {
                    activeList.RemoveAt(index);
                }
            }

            points.Sort((a, b) => a.y.CompareTo(b.y));

            return points;
        }
    }
}

File: gacha_test/lib/gacha/gacha_machine.dart
// lib/gacha/gacha_machine.dart

import 'dart:async';
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';
import '../utils/cancellation_token.dart';
import 'gacha_inside_ball.dart';

class GachaMachine {
  bool _active = true;

  // transform プロパティ（1.0 単位＝100px 換算）
  final ValueNotifier<Offset> _position =
      ValueNotifier<Offset>(const Offset(0, 5.0)); // 初期位置：画面上部（見えない位置）
  final ValueNotifier<Offset> _scale =
      ValueNotifier<Offset>(const Offset(1.0, 1.2));
  final ValueNotifier<double> _rotation = ValueNotifier<double>(0.0); // 単位：度
  final ValueNotifier<double> _knobRotation = ValueNotifier<double>(0.0);

  // 内部ボール群とボールに設定する色一覧
  final List<GachaInsideBall> _gachaBalls;
  final List<Color> _ballColors;

  // ガラス内部領域の中心（単位）
  final Offset _glassOrigin = const Offset(0.0, 3.0);
  static const double containerRadius = 1.0;

  GachaMachine({
    required List<GachaInsideBall> gachaBalls,
    required List<Color> ballColors,
  })  : _gachaBalls = gachaBalls,
        _ballColors = ballColors;

  /// 初期化処理
  void initialize() {
    _active = true;
    _position.value = const Offset(0, 5.0); // 画面上部（非表示領域）
    _rotation.value = 0.0;
    _scale.value = const Offset(1.0, 1.2);
    _knobRotation.value = 0.0;

    // Poisson Disk Sampling により内部ボールの位置を生成
    List<Offset> ballPositions =
        _generateBallPositions(_glassOrigin, containerRadius, 0.32, 64);
    final random = Random();
    for (int i = 0; i < _gachaBalls.length; i++) {
      final ball = _gachaBalls[i];
      ball.setPosition(ballPositions[i]);
      ball.setRandomAngle();
      // ランダムな色を設定
      Color randomColor = _ballColors[random.nextInt(_ballColors.length)];
      ball.setColor(randomColor);
    }
  }

  /// 落下アニメーション：位置とスケールの変更を並行実行
  Future<void> fallAsync({
    required Duration duration,
    required CancellationToken ct,
    required TickerProvider ticker,
  }) async {
    ct.throwIfCancellationRequested();

    final positionController =
        AnimationController(vsync: ticker, duration: duration);
    final scaleController =
        AnimationController(vsync: ticker, duration: duration);

    final positionTween =
        Tween<Offset>(begin: _position.value, end: const Offset(0, -2.24));
    final scaleTween =
        Tween<Offset>(begin: _scale.value, end: const Offset(1.0, 1.0));

    final positionCompleter = Completer<void>();
    final scaleCompleter = Completer<void>();

    positionController.addListener(() {
      _position.value = positionTween.evaluate(positionController);
      if (positionController.status == AnimationStatus.completed) {
        positionCompleter.complete();
      }
    });
    scaleController.addListener(() {
      _scale.value = scaleTween.evaluate(scaleController);
      if (scaleController.status == AnimationStatus.completed) {
        scaleCompleter.complete();
      }
    });

    positionController.forward();
    scaleController.forward();

    if (ct.isCancelled) {
      positionController.dispose();
      scaleController.dispose();
      throw CancellationException();
    }

    await Future.wait([positionCompleter.future, scaleCompleter.future]);
    positionController.dispose();
    scaleController.dispose();
  }

  /// 着地時のアニメーション（シェイクとリバウンドを並行実行）
  Future<void> landAsync({
    required CancellationToken ct,
    required TickerProvider ticker,
  }) async {
    await Future.wait([
      _shakingAnimation(duration: const Duration(milliseconds: 400), ct: ct),
      _reboundAnimation(ct: ct, ticker: ticker),
    ]);
  }

  Future<void> _shakingAnimation({
    required Duration duration,
    required CancellationToken ct,
  }) async {
    final original = _position.value;
    final stopwatch = Stopwatch()..start();
    final random = Random();
    while (stopwatch.elapsed < duration) {
      if (ct.isCancelled) break;
      // x軸方向に ±0.2 単位のランダムなオフセット
      double offsetX = (random.nextDouble() * 2 - 1) * 0.2;
      _position.value = original + Offset(offsetX, 0);
      await Future.delayed(const Duration(milliseconds: 16));
    }
    _position.value = original;
  }

  Future<void> _reboundAnimation({
    required CancellationToken ct,
    required TickerProvider ticker,
  }) async {
    await _animateScale(
      from: _scale.value,
      to: const Offset(1.1, 0.8),
      duration: const Duration(milliseconds: 125),
      ticker: ticker,
      ct: ct,
    );
    await _animateScale(
      from: _scale.value,
      to: const Offset(0.95, 1.05),
      duration: const Duration(milliseconds: 125),
      ticker: ticker,
      ct: ct,
    );
    await _animateScale(
      from: _scale.value,
      to: const Offset(1.0, 1.0),
      duration: const Duration(milliseconds: 125),
      ticker: ticker,
      ct: ct,
    );
  }

  Future<void> _animateScale({
    required Offset from,
    required Offset to,
    required Duration duration,
    required TickerProvider ticker,
    required CancellationToken ct,
  }) async {
    ct.throwIfCancellationRequested();
    final controller =
        AnimationController(vsync: ticker, duration: duration);
    final tween = Tween<Offset>(begin: from, end: to);
    final completer = Completer<void>();

    controller.addListener(() {
      _scale.value = tween.evaluate(controller);
      if (controller.status == AnimationStatus.completed) {
        completer.complete();
      }
    });

    controller.forward();
    if (ct.isCancelled) {
      controller.dispose();
      throw CancellationException();
    }

    await completer.future;
    controller.dispose();
  }

  /// ビートループ：一定間隔でスケールを変化させるループアニメーション
  Future<void> beatLoopAsync({
    required CancellationToken ct,
    required TickerProvider ticker,
  }) async {
    const beatScale = Offset(1.05, 0.95);
    try {
      while (!ct.isCancelled) {
        _scale.value = beatScale;
        await _animateScale(
          from: beatScale,
          to: const Offset(1.0, 1.0),
          duration: const Duration(milliseconds: 500),
          ticker: ticker,
          ct: ct,
        );
      }
    } finally {
      _scale.value = const Offset(1.0, 1.0);
    }
  }

  /// ノブの回転アニメーション
  Future<void> rotateKnobAsync({
    required Duration duration,
    required CancellationToken ct,
    required TickerProvider ticker,
  }) async {
    // 第1フェーズ：現在の回転から -90° へ（duration の 40%）
    await _animateKnobRotation(
      from: _knobRotation.value,
      to: -90.0,
      duration: Duration(milliseconds: (duration.inMilliseconds * 0.4).round()),
      ticker: ticker,
      ct: ct,
    );
    // 20% 分待機
    await Future.delayed(
      Duration(milliseconds: (duration.inMilliseconds * 0.2).round()),
      () {
        ct.throwIfCancellationRequested();
      },
    );
    // 第2フェーズ：-90° から -180° へ（duration の 40%）
    await _animateKnobRotation(
      from: _knobRotation.value,
      to: -180.0,
      duration: Duration(milliseconds: (duration.inMilliseconds * 0.4).round()),
      ticker: ticker,
      ct: ct,
    );
    // ノブの回転をリセット
    _knobRotation.value = 0.0;
  }

  Future<void> _animateKnobRotation({
    required double from,
    required double to,
    required Duration duration,
    required TickerProvider ticker,
    required CancellationToken ct,
  }) async {
    ct.throwIfCancellationRequested();
    final controller =
        AnimationController(vsync: ticker, duration: duration);
    final tween = Tween<double>(begin: from, end: to);
    final completer = Completer<void>();

    controller.addListener(() {
      _knobRotation.value = tween.evaluate(controller);
      if (controller.status == AnimationStatus.completed) {
        completer.complete();
      }
    });

    controller.forward();
    if (ct.isCancelled) {
      controller.dispose();
      throw CancellationException();
    }
    await completer.future;
    controller.dispose();
  }

  /// 抽選アニメーション：
  /// ・各内部ボールのバウンス＆回転ループを開始し、
  /// ・機体全体に軽い揺れ（回転シェイク）を与え、一定時間後にキャンセルします。
  Future<void> lotteryAsync({
    required Duration duration,
    required CancellationToken ct,
    required TickerProvider ticker,
  }) async {
    final lotteryCts = CancellationTokenSource.createLinked(ct);

    // 各内部ボールのバウンス＆回転ループ開始
    _gachaBalls.map((ball) {
      return ball.bounceLoopAsync(
        containerCenter: _glassOrigin,
        containerRadius: containerRadius,
        speed: 12.0,
        ct: lotteryCts.token,
        ticker: ticker,
      );
    }).toList();
    _gachaBalls.map((ball) {
      return ball.rotateLoopAsync(
        speed: 360.0,
        ct: lotteryCts.token,
        ticker: ticker,
      );
    }).toList();

    // 機体全体の回転を -2° に設定
    _rotation.value = -2.0;
    // 軽い回転シェイクを開始
    _startShakeRotation(ct: lotteryCts.token, ticker: ticker);

    // 指定時間待機後、抽選アニメーションを終了
    await Future.delayed(duration);
    lotteryCts.cancel();
  }

  Future<void> _startShakeRotation({
    required CancellationToken ct,
    required TickerProvider ticker,
  }) async {
    try {
      while (!ct.isCancelled) {
        await _animateRotation(
          from: -2.0,
          to: 2.0,
          duration: const Duration(milliseconds: 50),
          ticker: ticker,
          ct: ct,
        );
        await _animateRotation(
          from: 2.0,
          to: -2.0,
          duration: const Duration(milliseconds: 50),
          ticker: ticker,
          ct: ct,
        );
      }
    } catch (e) {
      // キャンセル例外は握りつぶす
    }
  }

  Future<void> _animateRotation({
    required double from,
    required double to,
    required Duration duration,
    required TickerProvider ticker,
    required CancellationToken ct,
  }) async {
    ct.throwIfCancellationRequested();
    final controller =
        AnimationController(vsync: ticker, duration: duration);
    final tween = Tween<double>(begin: from, end: to);
    final completer = Completer<void>();

    controller.addListener(() {
      _rotation.value = tween.evaluate(controller);
      if (controller.status == AnimationStatus.completed) {
        completer.complete();
      }
    });
    controller.forward();
    if (ct.isCancelled) {
      controller.dispose();
      throw CancellationException();
    }
    await completer.future;
    controller.dispose();
  }

  /// 非表示にする
  void hide() {
    _active = false;
  }

  /// Poisson Disk Sampling により、円形領域内で各点間が [minDistance] 以上離れている点群を生成し、
  /// Y 座標が低い順にソートしたリストを返します。
  List<Offset> _generateBallPositions(
      Offset center, double containerRadius, double minDistance, int desiredCount,
      {int k = 30, int? seed}) {
    final random = seed != null ? Random(seed) : Random();
    List<Offset> points = [];
    List<Offset> activeList = [];

    const int targetCircumferencePoints = 32;
    int validCircumferencePoints = 0;
    int attempts = 0;
    int maxAttempts = targetCircumferencePoints * 8;

    // 円周上のランダムな点を生成
    while (validCircumferencePoints < targetCircumferencePoints &&
        attempts < maxAttempts) {
      attempts++;
      double angle = random.nextDouble() * 2 * pi;
      Offset candidate =
          center + Offset(containerRadius * cos(angle), containerRadius * sin(angle));
      if (!points.every((p) => (candidate - p).distance >= minDistance)) continue;
      points.add(candidate);
      activeList.add(candidate);
      validCircumferencePoints++;
    }

    // Poisson Disk Sampling により内部の点を生成
    while (activeList.isNotEmpty && points.length < desiredCount) {
      int index = random.nextInt(activeList.length);
      Offset point = activeList[index];
      bool candidateAccepted = false;
      for (int i = 0; i < k; i++) {
        double angle = random.nextDouble() * 2 * pi;
        double distance = minDistance * (1 + random.nextDouble());
        Offset candidate = point + Offset(cos(angle), sin(angle)) * distance;
        if ((candidate - center).distance > containerRadius) continue;
        if (!points.every((p) => (candidate - p).distance >= minDistance)) continue;
        points.add(candidate);
        activeList.add(candidate);
        candidateAccepted = true;
        break;
      }
      if (!candidateAccepted) {
        activeList.removeAt(index);
      }
    }

    points.sort((a, b) => a.dy.compareTo(b.dy));
    return points;
  }

  /// Widget ツリーに組み込むためのビルドメソッド
  Widget buildWidget() {
    if (!_active) return const SizedBox.shrink();
    return ValueListenableBuilder<Offset>(
      valueListenable: _position,
      builder: (context, pos, child) {
        return ValueListenableBuilder<Offset>(
          valueListenable: _scale,
          builder: (context, scale, child) {
            return ValueListenableBuilder<double>(
              valueListenable: _rotation,
              builder: (context, rotation, child) {
                return Transform(
                  transform: Matrix4.identity()
                    ..translate(pos.dx * 100, pos.dy * 100)
                    ..rotateZ(rotation * pi / 180)
                    ..scale(scale.dx, scale.dy),
                  alignment: Alignment.center,
                  child: _buildMachineStack(),
                );
              },
            );
          },
        );
      },
    );
  }

  /// 機体の各レイヤーを重ねた Widget
  Widget _buildMachineStack() {
    return Stack(
      alignment: Alignment.bottomCenter,
      children: [
        // 背面レイヤー：GachaMachine.svg（304 x 456 px, origin: bottom）
        SvgPicture.asset(
          'assets/gacha/GachaMachine.svg',
          width: 304,
          height: 456,
          alignment: Alignment.bottomCenter,
        ),
        // 内部ボール（各ボールは ball.buildWidget() で描画）
        ..._gachaBalls.map((ball) => ball.buildWidget()),
        // ノブ：GachaMachineKnob.svg（80 x 80 px, origin: center, position: (0, 132)）
        ValueListenableBuilder<double>(
          valueListenable: _knobRotation,
          builder: (context, knobRot, child) {
            return Transform.translate(
              offset: const Offset(0, 132),
              child: Transform.rotate(
                angle: knobRot * pi / 180,
                child: SvgPicture.asset(
                  'assets/gacha/GachaMachineKnob.svg',
                  width: 80,
                  height: 80,
                  alignment: Alignment.center,
                ),
              ),
            );
          },
        ),
        // 前面レイヤー：GachaMachineFront.svg（304 x 456 px, origin: bottom）
        SvgPicture.asset(
          'assets/gacha/GachaMachineFront.svg',
          width: 304,
          height: 456,
          alignment: Alignment.bottomCenter,
        ),
      ],
    );
  }
}
